### PR TITLE
gh-105535 Document potential performance trap during enum creation

### PR DESCRIPTION
**Pull request by nicky-eng**
Fix issue #105535 

Document potential performance trap during enum creation. 

* Issue: gh-105535

----
:books: Documentation preview :books:: https://cpython-previews--107119.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/enum.rst [hunk 1]
```diff
@@ -267,6 +267,10 @@ Data Types
          appropriate value will be chosen for you.  See :class:`auto` for the
          details.
 
+         If the value is not hashable, the enum creation will take quadratic time
+         relative to the number of members.
```
Comment (evildmp) [inline Doc/library/enum.rst]: I think I might express this as something like:

> ... [be warned that] the time required for enum creation will grow quadratically with the number of members.

Comment (nicky-eng) [reply Doc/library/enum.rst]: Updated wording to include this as a warning.

**Review by evildmp — APPROVED** [review]

File: Doc/library/enum.rst [hunk 2]
```diff
@@ -267,6 +267,10 @@ Data Types
          appropriate value will be chosen for you.  See :class:`auto` for the
          details.
 
+         If the values are not hashable, be warned that the time required for enum
+         creation will grow quadratically with the number of members.
```
Comment (ethanfurman) [inline Doc/library/enum.rst]:
```suggestion
         While mutable/unhashable values, such as :class:`dict` or :class:`list`, can
         be used, they will have a quadratic performance impact relative to the total
         number of mutable/unhashable values in the enum.
```

File: Doc/library/enum.rst [hunk 3]
```diff
@@ -267,6 +267,11 @@ Data Types
          appropriate value will be chosen for you.  See :class:`auto` for the
          details.
 
+         While mutable/unhashable values, such as :class:`dict` or :class:`list`, can
```
Comment (evildmp) [inline Doc/library/enum.rst]: I'd prefer to see the two commas on this line discarded, they introduce too many pauses.

**Review by evildmp — CHANGES_REQUESTED** [review]

File: Doc/library/enum.rst (hunk 3 above)
Comment (ethanfurman) [reply Doc/library/enum.rst]: Pauses aren't necessarily bad, and I don't think parentheses are warranted here.

File: Doc/library/enum.rst [hunk 4]
```diff
@@ -267,6 +267,11 @@ Data Types
...
+         be used, they will have a quadratic performance impact relative to the total
+         number of mutable/unhashable values in the enum.
```
Comment (ethanfurman) [inline Doc/library/enum.rst]:
```suggestion
         While mutable/unhashable values, such as :class:`dict`, :class:`list` or
         a mutable :class:`dataclass`, can be used, they will have a quadratic
         performance impact relative to the total number of mutable/unhashable
         values in the enum.
```

**Comment by ethanfurman** [comment]
Thoughts on making this an actual warning box?

File: Doc/library/enum.rst [hunk 5]
```diff
@@ -267,6 +267,12 @@ Data Types
          appropriate value will be chosen for you.  See :class:`auto` for the
          details.
 
+         While mutable/unhashable values, such as :class:`dict`, :class:`list` or
+         a mutable :class:`dataclass`, can be used, they will have a quadratic
```
Comment (hugovk) [inline Doc/library/enum.rst]: The ```:class:`dataclass` ``` reference can't be found, can this be fixed?

File: Doc/library/enum.rst [hunk 6]
```diff
@@ -267,6 +267,12 @@ Data Types
...
+         performance impact relative to the total number of mutable/unhashable
+         values in the enum.
```
Comment (ethanfurman) [inline Doc/library/enum.rst]:
```suggestion
         a mutable :class:`~dataclasses.dataclass`, can be used, they will have a
         quadratic performance impact relative to the total number of
         mutable/unhashable values in the enum.
```

**Review by evildmp — APPROVED** [review]

File: Doc/library/enum.rst [hunk 7]
```diff
@@ -267,6 +267,12 @@ Data Types
          appropriate value will be chosen for you.  See :class:`auto` for the
          details.
 
+         While mutable/unhashable values, such as :class:`dict`, :class:`list` or
+         a mutable :class:`~dataclasses.dataclass`, can be used, they will have a
+         quadratic performance impact relative to the total number of
+         mutable/unhashable values in the enum.
```
Comment (ethanfurman) [inline Doc/library/enum.rst]:
```suggestion
         quadratic performance impact during creation relative to the
         total number of mutable/unhashable values in the enum.
```

File: Doc/library/enum.rst [hunk 8]
```diff
@@ -278,21 +278,20 @@ Data Types
 
          >>> Color.RED.value
          1
-
+         
+      Value of the member, can be set in :meth:`~object.__new__`.
+
```
Comment (ethanfurman) [inline Doc/library/enum.rst]:
```suggestion

      Value of the member, can be set in :meth:`~object.__new__`.

```

File: Doc/library/enum.rst [hunk 9]
```diff
@@ -279,20 +279,19 @@ Data Types
...
-
-      Name of the member.
-
-   .. attribute:: Enum._value_
-
-      Value of the member, can be set in :meth:`~object.__new__`.
+         While mutable/unhashable values, such as :class:`dict`, :class:`list` or
+         a mutable :class:`~dataclasses.dataclass`, can be used, they will have a
+         quadratic performance impact during creation relative to the
+         total number of mutable/unhashable values in the enum.
```
Comment (hugovk) [inline Doc/library/enum.rst:295]: To fix the CI:

I think these were accidentally removed when merging in `main` (they were recently added in https://github.com/python/cpython/pull/114696).

```suggestion

.. attribute:: Enum._name_

      Name of the member.

   .. attribute:: Enum._value_

      Value of the member, can be set in :meth:`~object.__new__`.

```

File: Doc/library/enum.rst [hunk 10]
```diff
@@ -279,14 +279,21 @@ Data Types
...
          appropriate value will be chosen for you.  See :class:`auto` for the
          details.
 
-   .. attribute:: Enum._name_
+         While mutable/unhashable values, such as :class:`dict`, :class:`list` or
+         a mutable :class:`~dataclasses.dataclass`, can be used, they will have a
+         quadratic performance impact during creation relative to the
+         total number of mutable/unhashable values in the enum.
+
+.. attribute:: Enum._name_
```
Comment (hugovk) [inline Doc/library/enum.rst]: Oops, my copy/paste missed some whitespace:

```suggestion
   .. attribute:: Enum._name_
```

**Review by hugovk — APPROVED** [review]

**Pull request merged**